### ISSUE TITLE
Title of a Movie not shown in the player

### ISSUE THREAD

**Issue by elgatito** [scakemyer/plugin.video.quasar#739]
On Kodi17 (not tried with previous versions) when you play episode from Kodi library - in the left upper corner you see the title. When you play Movie you just see "play" text in that corner.

## Expected Behavior

See the movie title as well.

## Steps to Reproduce (for bugs)

1.
2.
3.
4.

## Your Environment

* Version used: 0.9.72
* Environment name and version (e.g. Kodi 15/16/17, SPMC, OSMC, OpenELEC): Openelec
* Operating System and version (Android 4/5/6, Windows 7/8/10, macOS 10.11, Ubuntu 16.10): Windows
* Link to Kodi debug log:
<!--- See http://kodi.wiki/view/Log_file/Easy or paste to http://paste.ubuntu.com directly

**Comment by scakemyer** [comment]
I see the movie titles just fine here, you'll have to be more specific about the context, the steps you're taking, which skin you're using, etc.

**Comment by elgatito** [comment]
Estuary, everything is default.
Movie added to the library from Quasar. Then play it from library, Quasar gets stream automatically and player is started. In the left upper corner I see no title, but see "play" text.

**Comment by elgatito** [comment]
![screen2](https://cloud.githubusercontent.com/assets/1247242/23012965/4fa7744a-f431-11e6-96aa-d2e6847686eb.png)
Okay, here is the screen. Now you can see.

strm content:
```
plugin://plugin.video.quasar/movie/284052/links

```

So "play" in issue description is because on another Kodi stream was added with "select stream automatically", which is written by ```library.go``` at ```writeMovieStrm```.

**Comment by scakemyer** [comment]
Playing from library or Quasar itself, I get proper titles all around. It's normal to get "play" or "links" when starting playback from the WebUI, or from other external sources, since we simply don't have any infotags in those cases; we're playing raw torrent files, and all that Kodi sees is the `plugin://plugin.video.quasar` URI, takes the last part as the "filename" and displays that.

**Comment by elgatito** [comment]
![image](https://cloud.githubusercontent.com/assets/1247242/23014931/aa4e02bc-f439-11e6-9216-5bb4fc2f3184.png)
I've been playing with library to get it.
So you can see I get a title when playing from a library. 
That is made by including NFO file near the strm file. It uses meta information from it automatically. But that should be done when creating strm files, before library gets updated. 

From http://kodi.wiki/view/NFO_files/movies :
```
<movie>
        <title>Hellboy - Uncut Extended</title>
        <set>Hellboy Collection</set>
        <sorttitle>Hellboy 1</sorttitle>
</movie>
http://www.themoviedb.org/movie/1487
```

We can store link to the moviedb with a title, which is already localized in Quasar, according to Kodi locale.

**Comment by elgatito** [comment]
![image](https://cloud.githubusercontent.com/assets/1247242/23015496/b57f5bfc-f43b-11e6-94b1-10e43a945d3d.png)
Awesome! 
Placing nfo file with this content:
```
http://www.themoviedb.org/movie/1487
```

The link is for Hellboy (Which title in russian is visible on the screen). It made themoviedb scraper link that item in the library properly and Kodi uses information when playing the file.

So the link can we written, depending which scraper is selected in Quasar settings.

**Comment by scakemyer** [comment]
Ok so basically you weren't using Kodi's scraping functionality in the first place for your library. Well that explains it. Instead of making .nfo files, you could just [set your content](http://kodi.wiki/view/Set_content_and_scan) per library folder, set the scraper to use in Kodi, and enjoy.

**Comment by elgatito** [comment]
I did use it actually. That directory was set with TheMovieDB, I did see posters, title localized and so on.
But it did not work when Kodi launched strm play.

**Comment by elgatito** [comment]
This is a log cut when updating library with created by Quasar entry in Movies:
```
10:12:31.131 T:5124   DEBUG: VideoInfoScanner: Scanning dir 'C:\Incoming\Movies\Doctor Strange (2016)\' as not in the database
10:12:31.135 T:5124   DEBUG: VideoInfoScanner: No NFO file found. Using title search for 'C:\Incoming\Movies\Doctor Strange (2016)\Doctor Strange (2016).strm'
10:12:31.136 T:5124   DEBUG: ADDON::CScraper::FindMovie: Searching for 'Doctor Strange' using The Movie Database scraper (path: 'C:\Program Files (x86)\Kodi\addons\metadata.themoviedb.org', content: 'movies', version: '3.9.2')
10:12:31.251 T:5124   DEBUG: scraper: CreateSearchUrl returned <url>http://api.tmdb.org/3/search/movie?api_key=f7f51775877e0bb6703520952b3c7840&amp;query=Doctor%20Strange&amp;year=2016&amp;language=ru</url>
10:12:31.273 T:5124   DEBUG: CurlFile::Open(0D34CF98) http://api.tmdb.org/3/search/movie?api_key=f7f51775877e0bb6703520952b3c7840&query=Doctor%20Strange&year=2016&language=ru
10:12:31.273 T:5124    INFO: XCURL::DllLibCurlGlobal::easy_aquire - Created session to http://api.tmdb.org
10:12:31.537 T:5124   DEBUG: CScraperUrl::Get: Using "UTF-8" charset for "http://api.tmdb.org/3/search/movie?api_key=f7f51775877e0bb6703520952b3c7840&query=Doctor%20Strange&year=2016&language=ru"
10:12:31.550 T:5124   DEBUG: scraper: GetSearchResults returned <results><entity><title>Доктор Стрэндж</title><id>284052</id><year>2016</year><url cache="tmdb-ru-284052.json">http://api.tmdb.org/3/movie/284052?api_key=f7f51775877e0bb6703520952b3c7840&amp;language=ru</url></entity><entity><title>Doctor Strange</title><id>284052</id><year>2016</year><url cache="tmdb-ru-284052.json">http://api.tmdb.org/3/movie/284052?api_key=f7f51775877e0bb6703520952b3c7840&amp;language=ru</url></entity></results>

```

If I run that movie from a library - I see "play" or "links" title in the corner.
If I add nfo file with the same link to themoviedb - I see the title while playing strm.

**Comment by elgatito** [comment]
Probably Kodi expecting something from a stream. 
When nfo is added is acts like a forced information source.

It would be a few lines change in library.go to write "http://www.themoviedb.org/movie/{tmdbId}" with the same name, same directory, but ".nfo" extension.

**Comment by scakemyer** [comment]
Ok so either this is Kodi thing, or there's yet a few other unicode/encoding bugs with regards to setting infolabels, and using nfo files bypasses that. I don't like the idea of creating .nfo files along with .strm files since it forces users use one scraper vs another. It's one thing to use TMDB all around Quasar but we shouldn't prevent users from choosing differently for their own library.

**Comment by elgatito** [comment]
Without NFO there is not title in the player.

NFO can have a link to TMDB (if scraper not installed - will be skipped), or just a single <title> tag with localized title. If second - all other information will be used from selected scraper, and only title from the NFO file.

Do you see the title ever for movies, launched from a library?

**Comment by scakemyer** [comment]
I get that, and yes, always.

**Comment by scakemyer** [comment]
If anything it probably has to do with `xbmc.getInfoLabel` returning strings instead of unicode, then setting that in `xbmcgui.ListItem.setInfo` in `navigation.py`...

**Comment by elgatito** [comment]
It's strange... 
- I have changed Kodi locale to English. 
- Quasar showing items in English. FIne
- Adding to the library
- Library configured with TMDB and language is the same - russian.
- Movie appeared in the library with russian information and so on.
- Playing from library - I see full information in the left upper corner in russian just fine. 
- ??

**Comment by elgatito** [comment]
http://paste.ubuntu.com/24006459/
First is opening one entry with english, second - same entry with russian.
Maybe you can see anything.

**Comment by scakemyer** [comment]
Nah nothing in there..

**Comment by elgatito** [comment]
I have found the problem... 
When launching play from main window in Estuary - it gets into navigation.py and then ListItem is empty at all. If launching from list of films - ListItem is populated.
Crazy...

**Comment by KillerJoeBR** [comment]
Yeah I noticed this behavior sometime ago.  I thought it was something that couldn't be fixed. So I didn't report. But there is this limitation with Quasar while playing a movie from the widgets on kodi home page.  If you play from inside the library then everything works fine.

**Comment by elgatito** [comment]
Huh.. Another thing caused by this - resume state is not saved as well. I guess it does not know which item is actually playing. I will try to find the way to resolve this. Don't like this.

**Comment by scakemyer** [comment]
I have an idea for populating the InfoLabels when the ListItem is empty. Since we have the TMDB id in the plugin URI, we can query the daemon for the info and use that instead. Of course it'd be a lot simpler if Kodi actually played nice and returned something to `getInfoLabel` in the first place, regardless of if we're in the library or on the home page, but that's Kodi... Saving the resume state and watched status without a Kodi item ID is going to be a serious pita though.

**Comment by elgatito** [comment]
Yeah, that will only help to show information. Which is already in Kodi.. I will try to find a way out of this.

**Comment by scakemyer** [comment]
The info being "already in Kodi" is useless to us if Kodi isn't returning it in the first place when we ask for it, on items that clearly have it too. There's no easy way out of it mate, you'll find plenty of issues related to this one, to which I never had an answer, and querying the daemon using the TMDB id in the URI is the only thing we can do to be honest. A similar solution would be to query it using that TMDB id, but instead of getting TMDB movie/episode info, finding the Kodi item ID, querying Kodi right back for its info, and returning that to the add-on. It's loopy but it should work.

**Comment by elgatito** [comment]
Is it possible to write originaltitle here in the link as well? - scakemyer/quasar@acc1e5e
I've been searching for a good answer for this issue, but ...

We can't query Kodi's Json-RPC to search for TMDB id directly. 
We can search title, but not original title, only the one that is listed in Kodi library.
We can query by filename (strm file), then yes, we will need to query TMDB for name and year, OR, we can include that in the link inside of strm file.
It appears that once the stream link is loaded - Kodi loses information about the file, that triggered the stream, so we can add filename into the url, then we can just query json-rpc for details and set it to ListItem.

Tried many different ways to avoid searching for movies, but nothing worked, every method lacks something.

**Comment by scakemyer** [comment]
Don't waste time on this, should be pretty trivial to query back the daemon with the TMDB id to get the info, I'll get around to it soon enough.

**Comment by elgatito** [comment]
It's digging me! The thing, you know it should have something, trivial and simple, but you can't get it.

**Comment by elgatito** [comment]
Issue is getting wider. 
When you open a series from main menu, you see a list of episodes, when you open one - information is passed to navigation.py, but when Kodi runs "next episode" - no information (because listitem is not populated). So no "seen" or "resume" data is updated.

**Comment by scakemyer** [comment]
That doesn't depend on infotags, but the `Player.OnPlay` notifications. I'll check what notification we do get and what's causing that issue, but I don't think it's related.

**Comment by elgatito** [comment]
Every time I see "links/play" on the player - resume/watched data not saved. I thought it's because item not resolved in navigation.py.

**Comment by scakemyer** [comment]
Nah, what's in `navigation.py` has nothing to do with resume/watched status getting saved or not.

**Comment by elgatito** [comment]
So what is the place in the code, where you set link to the library item and so on?

**Comment by scakemyer** [comment]
We don't "set" a link to a library item anywhere. We do load (and reload when necessary) the Kodi library infos we need for duplicate checks, but even that has nothing to do with setting resume/watched status. That's all from simple `Player.OnStop` notifications in `api/library.go`, which include the Kodi library item ID, and we just use that directly to set an episode's or movie's watched status. I haven't looked into that secondary issue yet, but my guess is that when Kodi goes to play the next episode, we just don't receive any OnStop notification, in which case I'll need to look into what notification we do get as I mentioned.

**Comment by elgatito** [comment]
When I tried to find any usable information in InfoLabels of Kodi I've tried many different information containers like Player.*. Container.*. System.*. If ListItem is populated - all those containers return proper information, otherwise - they are empty. 
I have an idea that if we define information in navigation.py, then Player will send it to onPlay and it will be used in proper way.
if you add debug messages to On* functions ( I can't rebuild daemon on my system), I will try that idea.

**Comment by scakemyer** [comment]
That's not how it works, `OnPlay` and `OnStop` notifications don't contain anything more than a Kodi library item ID and a type, doesn't matter whether there's a ListItem with information or not. I know this is bothering you and I'll get around to fixing it, but again, you're still wasting your time on this right now (and mine).

**Comment by elgatito** [comment]
Update on this one.

I have:
- recompiled binary to show incoming Notification messages, coming from Kodi (Notification() in library.go)
- modified notification.py (which currently creates ListItem and transfer properties from Kodi's ListItem to plugin's ListItem).

The change in navigation.py:
```

            if not _infoLabels['DBID']:
                _infoLabels = eval("{'Plot': '', 'PlotOutline': '', 'VideoCodec': '', 'Title': \"Pete's Dragon\", 'Tagline': 'Some secrets are too big to keep.', 'TVShowTitle': '', 'Writer': 'David Lowery / Toby Halbrooks', 'Votes': '377', 'FileNameAndPath': \"C:\\Incoming\\Movies\\Pete's Dragon (2016)\\Pete's Dragon (2016).strm\", 'DBID': '2', 'Season': '', 'Mediatype': 'movie', 'Director': 'David Lowery', 'Studio': 'Walt Disney Pictures', 'Genre': '', 'Path': \"C:\\Incoming\\Movies\\Pete's Dragon (2016)\", 'VideoResolution': '', 'VideoAspect': '', 'Episode': '', 'Premiered': '10.08.2016', 'Trailer': 'plugin://plugin.video.youtube/?action=play_video&videoid=7aNEwtO-Jx4', 'Rating': '6.1', 'Code': 'tt2788732', 'DBTYPE': 'movie', 'MPAA': 'Rated 6+', 'FileName': \"Pete's Dragon (2016).strm\", 'OriginalTitle': \"Pete's Dragon\", 'IMDBNumber': 'tt2788732', 'Year': '2016', 'FolderPath': 'videodb://movies/titles/2'}")
            infoLabels = {}
            for key, value in _infoLabels.iteritems():
                if value:
                    infoLabels[key] = value
                    xbmcplugin.setProperty(HANDLE, key, value)

```

So when ListItem is empty, I set it with default value (but real, in this case movie's DBID is "2").
Then I'm setting plugin's properties, so anytime requests come from plugin - it reads this property.
```
xbmcplugin.setProperty(HANDLE, key, value)
```

I have checked responses coming from Kodi. With this change it passes all the IDs to onPlay/onSeek/onUpdate and other notifications. So it looks like it's also using ListItem, which is null, by default.

So what is needed for this:
From daemon:
- We have only an url, so only TMDB id available and an item type (episode/movie). That is why we need to perform TMDB resolve, 
- Then find DBID by querying JSON-API (see http://kodi.wiki/view/JSON-RPC_API/v6#VideoLibrary.GetMovies action with http://kodi.wiki/view/JSON-RPC_API/v6#List.Filter.Fields.Movies with "filename" filter field to find by filename (filename is what we generate by the daemon when we write it for the library addition))
- Then do http://kodi.wiki/view/JSON-RPC_API/v6#VideoLibrary.GetMovieDetails to fetch all the information
- Pass all that to navigation.py

From python:
- Call daemon to resolve item
- Parse result and set it to setProperty()

That should be all we need.
I can help with it, but not sure what is the better way of communicating between python<>go.

**Comment by scakemyer** [comment]
You're getting close ;) A little overcomplicated in some places and not enough in others, but still. And we'll now be able to use the torrent DB implemented for #440 to get infotags for plain torrents also.

So on the daemon side, what's mostly needed is just to expose a special GetMovie / GetEpisode. We already have TMDB / DBID mappings from library integration too, so no need to query Kodi again with a filename filter and all, neither do we need to do `VideoLibrary.GetMovieDetails`, we already have all that. But you have the right idea overall :)
